### ISSUE TITLE
AWS control plane kubernetes version change

### ISSUE THREAD

**Issue by coffeepac** [samsung-cnct/k2#219]
what is the manual process for rolling forward or rolling backwards the version of kubernetes running on an aws control plane node?

this ticket should produce a set of command line steps to effect both up and down version changes.  that list of steps may also include a number of pauses to wait for nodes to go up and down.  I expect this to use 'drain' and then destroy the running node.  

This should also include any considerations for which versions of the control plane can speak to each other across versions.  this may also incur a cluster 'freeze' as the control plane is replaced.  this does not include etcd version changes.

**Comment by coffeepac** [comment]
we talked about this during our 1:1 and the easiest path may be to say 'while the control plane is upgrading, no changes to the system can be made.'  And then you can: 
- upgrade the ASG configs
- destroy all of these nodes in kubernetes/AWS
- let the ASG recreate them

kubernetes created pods can carry on just fine without a control plane (provided they don't die).  And all of the state of kubernetes is stored in etcd, not on the control plane nodes.  I'm okay with saying that a quiet period has to happen to upgrade the nodes but I'm not sure if everything will work.  Give it a shot!

**Comment by DStorck** [comment]
https://github.com/kubernetes/kubernetes/blob/release-1.4/docs/design/versioning.md#upgrades

**Comment by DStorck** [comment]
[Coreos article on k8s upgrading kubernetes components](https://coreos.com/kubernetes/docs/latest/kubernetes-upgrade.html)
Relevant section is towards the bottom of the page

**Comment by DStorck** [comment]
First attempt has been to use the bash method described in #218.  When all master nodes were torn down at once , the cluster became non-functional (app that had been running became unreachable and kubectl commands all returned  `Unable to connect to the server: EOF`.) 
Second attempt was to tear down one at a time, and this has yielded better results.  It took a little under 10 minutes to recreate a single node (not including the time it took for `k2cli cluster up` to reconfigure the version. 
Basic command was:
```k drain ip-10-0-87-80.us-west-2.compute.internal --force --ignore-daemonsets ; k delete node ip-10-0-87-80.us-west-2.compute.internal ; aws ec2 describe-instances --filters "Name=private-dns-name, Values=ip-10-0-87-80.us-west-2.compute.internal" > node.json; inst=$(cat node.json | jq -r '.Reservations[].Instances[].InstanceId'); aws ec2 terminate-instances --instance-ids $inst```

**Comment by DStorck** [comment]
Was able to bring upgrade all master nodes one at a time with this : 

```
#! /bin/bash

kubectl get no | awk '/Sched/{print $1}' > master_nodes.txt

cat master_nodes.txt | while read line; do
  kubectl drain $line --force --ignore-daemonsets; kubectl delete node $line;
  inst="$(aws ec2 describe-instances --filters "Name=private-dns-name, Values=$line" | jq -r '.Reservations[].Instances[].InstanceId')";
  aws ec2 terminate-instances --instance-ids $inst;

  number_nodes="$(kubectl get nodes | awk '/Sched/{print $1}' | wc -l)"

  until [ $number_nodes -eq 3 ]; do
    number_nodes="$(kubectl get nodes | awk '/Sched/{print $1}' | wc -l)"
    echo "waiting for nodes to be respawned"
    sleep 120
  done

  echo 'Master nodes have all been upgraded.'

done
```

**Comment by DStorck** [comment]
I'm sure there is a snazzier way to grab the number of master nodes out of the config file but this was just to make sure it worked alright. function now, fanciness later.

**Comment by DStorck** [comment]
Here is the output from the successful attempt: 
```node "ip-10-0-106-181.us-west-2.compute.internal" already cordoned
WARNING: Ignoring DaemonSet-managed pods: fluentd-daemon-2qq64, canal-node-lt9d3; Deleting pods not managed by ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet: kube-apiserver-ip-10-0-106-181.us-west-2.compute.internal, kube-controller-manager-ip-10-0-106-181.us-west-2.compute.internal, kube-proxy-ip-10-0-106-181.us-west-2.compute.internal, kube-scheduler-ip-10-0-106-181.us-west-2.compute.internal
node "ip-10-0-106-181.us-west-2.compute.internal" drained
node "ip-10-0-106-181.us-west-2.compute.internal" deleted
{
    "TerminatingInstances": [
        {
            "InstanceId": "i-026a54de230c5b917",
            "CurrentState": {
                "Code": 32,
                "Name": "shutting-down"
            },
            "PreviousState": {
                "Code": 16,
                "Name": "running"
            }
        }
    ]
}
Unable to connect to the server: EOF
waiting for nodes to be respawned
waiting for nodes to be respawned

node "ip-10-0-146-131.us-west-2.compute.internal" already cordoned
WARNING: Ignoring DaemonSet-managed pods: fluentd-daemon-bl6rc, canal-node-qv9kv; Deleting pods not managed by ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet: kube-apiserver-ip-10-0-146-131.us-west-2.compute.internal, kube-controller-manager-ip-10-0-146-131.us-west-2.compute.internal, kube-proxy-ip-10-0-146-131.us-west-2.compute.internal, kube-scheduler-ip-10-0-146-131.us-west-2.compute.internal
node "ip-10-0-146-131.us-west-2.compute.internal" drained
node "ip-10-0-146-131.us-west-2.compute.internal" deleted
{
    "TerminatingInstances": [
        {
            "InstanceId": "i-0a142cd723afc0494",
            "CurrentState": {
                "Code": 32,
                "Name": "shutting-down"
            },
            "PreviousState": {
                "Code": 16,
                "Name": "running"
            }
        }
    ]
}
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned

node "ip-10-0-43-12.us-west-2.compute.internal" already cordoned
WARNING: Ignoring DaemonSet-managed pods: fluentd-daemon-zl1cs, canal-node-l52zp; Deleting pods not managed by ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet: kube-apiserver-ip-10-0-43-12.us-west-2.compute.internal, kube-controller-manager-ip-10-0-43-12.us-west-2.compute.internal, kube-proxy-ip-10-0-43-12.us-west-2.compute.internal, kube-scheduler-ip-10-0-43-12.us-west-2.compute.internal
node "ip-10-0-43-12.us-west-2.compute.internal" drained
node "ip-10-0-43-12.us-west-2.compute.internal" deleted
{
    "TerminatingInstances": [
        {
            "InstanceId": "i-05cd59a64f66bcc93",
            "CurrentState": {
                "Code": 32,
                "Name": "shutting-down"
            },
            "PreviousState": {
                "Code": 16,
                "Name": "running"
            }
        }
    ]
}
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
all done
```

**Comment by DStorck** [comment]
Second runthrough initially produced this : 
```
node "ip-10-0-127-132.us-west-2.compute.internal" already cordoned
WARNING: Ignoring DaemonSet-managed pods: fluentd-daemon-ngnx3, canal-node-j1dpq; Deleting pods not managed by ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet: kube-apiserver-ip-10-0-127-132.us-west-2.compute.internal, kube-controller-manager-ip-10-0-127-132.us-west-2.compute.internal, kube-proxy-ip-10-0-127-132.us-west-2.compute.internal, kube-scheduler-ip-10-0-127-132.us-west-2.compute.internal
node "ip-10-0-127-132.us-west-2.compute.internal" drained
node "ip-10-0-127-132.us-west-2.compute.internal" deleted
{
    "TerminatingInstances": [
        {
            "InstanceId": "i-0653cd918d35353a4",
            "CurrentState": {
                "Code": 32,
                "Name": "shutting-down"
            },
            "PreviousState": {
                "Code": 16,
                "Name": "running"
            }
        }
    ]
}
Unable to connect to the server: EOF
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
Master nodes have all been upgraded.
node "ip-10-0-197-24.us-west-2.compute.internal" already cordoned
WARNING: Ignoring DaemonSet-managed pods: fluentd-daemon-mghlr, canal-node-0ggp1; Deleting pods not managed by ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet: kube-apiserver-ip-10-0-197-24.us-west-2.compute.internal, kube-controller-manager-ip-10-0-197-24.us-west-2.compute.internal, kube-proxy-ip-10-0-197-24.us-west-2.compute.internal, kube-scheduler-ip-10-0-197-24.us-west-2.compute.internal
node "ip-10-0-197-24.us-west-2.compute.internal" drained
node "ip-10-0-197-24.us-west-2.compute.internal" deleted

An error occurred (RequestLimitExceeded) when calling the DescribeInstances operation (reached max retries: 4): Request limit exceeded.

An error occurred (InvalidParameterCombination) when calling the TerminateInstances operation: No instances specified
waiting for nodes to be respawned
waiting for nodes to be respawned
waiting for nodes to be respawned
```

**Comment by DStorck** [comment]
To deal with RequestLimitExceeded issue, I put in a check for $inst 
```
#! /bin/bash

kubectl get no | awk '/Sched/{print $1}' > master_nodes.txt

cat master_nodes.txt | while read line; do
  kubectl drain $line --force --ignore-daemonsets; kubectl delete node $line;
  inst="$(aws ec2 describe-instances --filters "Name=private-dns-name, Values=$line" | jq -r '.Reservations[].Instances[].InstanceId')";

  until [[ $inst == i-* ]]; do
    sleep 60
    inst="$(aws ec2 describe-instances --filters "Name=private-dns-name, Values=$line" | jq -r '.Reservations[].Instances[].InstanceId')"
  done

  aws ec2 terminate-instances --instance-ids $inst;

  number_nodes="$(kubectl get nodes | awk '/Sched/{print $1}' | wc -l)"

  until [ $number_nodes -eq 3 ]; do
    number_nodes="$(kubectl get nodes | awk '/Sched/{print $1}' | wc -l)"
    echo "waiting for nodes to be respawned"
    sleep 120
  done

  echo 'Master nodes have all been upgraded.'
done
```

**Comment by DStorck** [comment]
To Review: 
try this out. You will need jq and aws-cli installed. The general idea is to change your k8s version in your config.yaml and then run k2cli up. After that you can run this script. Upgrade info about how many versions to step up or are down are referenced above in [this comment](https://github.com/samsung-cnct/k2/issues/219#issuecomment-286591123)

**Comment by coffeepac** [comment]
forced hand raise to review this isssss....   @venezia !  the big winner!

**Comment by guineveresaenger** [comment]
Tried running the script. For some reason, kubectl wasn't working without sourcing the entire filepath to admin.config, and this made it so the script didn't run. This might be my mistake - I used k2 from scratch to bring up the cluster, where in the past I had used k2cli.

When adjusting the script for this, the  process got hung on the aws request for the node name, after draining and deleting from k8s.
